src/providers/hana/qgshanaproviderconnection.cpp

```diff
@@ -0,0 +1,354 @@
+  {
+    throw QgsProviderConnectionException( QObject::tr( "An error occurred while creating the vector layer: %1" ).arg( errCause ) );
+  }
+}
+
+QString QgsHanaProviderConnection::tableUri( const QString &schema, const QString &name ) const
+{
+  const auto tableInfo { table( schema, name ) };
+
+  QgsDataSourceUri dsUri( uri() );
+  dsUri.setTable( name );
+  dsUri.setSchema( schema );
+  dsUri.setGeometryColumn( tableInfo .geometryColumn() );
+  return dsUri.uri( false );
+}
+
+void QgsHanaProviderConnection::dropVectorTable( const QString &schema, const QString &name ) const
+{
+  checkCapability( Capability::DropVectorTable );
+  dropTable( schema, name );
+}
+
+void QgsHanaProviderConnection::renameTable( const QString &schema, const QString &name, const QString &newName ) const
+{
+  executeSqlStatement( QStringLiteral( "RENAME TABLE %1.%2 TO %1.%3" )
+                       .arg( QgsHanaUtils::quotedIdentifier( schema ) )
+                       .arg( QgsHanaUtils::quotedIdentifier( name ) )
+                       .arg( QgsHanaUtils::quotedIdentifier( newName ) ) );
+}
+
+void QgsHanaProviderConnection::renameVectorTable( const QString &schema, const QString &name, const QString &newName ) const
+{
+  checkCapability( Capability::RenameVectorTable );
+  renameTable( schema, name, newName );
+}
+
+void QgsHanaProviderConnection::createSchema( const QString &name ) const
+{
+  checkCapability( Capability::CreateSchema );
+  executeSqlStatement( QStringLiteral( "CREATE SCHEMA %1" )
+                       .arg( QgsHanaUtils::quotedIdentifier( name ) ) );
+}
+
+void QgsHanaProviderConnection::dropSchema( const QString &name,  bool force ) const
+{
+  checkCapability( Capability::DropSchema );
+  executeSqlStatement( QStringLiteral( "DROP SCHEMA %1 %2" )
+                       .arg( QgsHanaUtils::quotedIdentifier( name ) )
+                       .arg( force ? QStringLiteral( "CASCADE" ) : QString() ) );
+}
+
+QList<QVariantList> QgsHanaProviderConnection::executeSql( const QString &sql ) const
+{
+  checkCapability( Capability::ExecuteSql );
+  if ( sql.trimmed().startsWith( "SELECT", Qt::CaseSensitivity::CaseInsensitive ) )
+    return executeSqlQuery( sql );
+  else
+  {
+    executeSqlStatement( sql );
+    return QList<QVariantList>();
+  }
+}
+
+QList<QVariantList> QgsHanaProviderConnection::executeSqlQuery( const QString &sql ) const
+{
+  const QgsDataSourceUri dsUri { uri() };
+  QList<QVariantList> results;
+  QgsHanaConnectionRef conn( dsUri );
+  if ( conn.isNull() )
+    throw QgsProviderConnectionException( QObject::tr( "Connection failed: %1" ).arg( uri() ) );
+
+  try
+  {
+    ConnectionRef connRef = conn->getNativeRef();
+    StatementRef stmt = connRef->createStatement();
+    QgsHanaResultSetRef resultSet = QgsHanaResultSet::create( stmt, sql );
+    ResultSetMetaDataRef metadata = resultSet->getMetadata();
+    const unsigned short nColumns = metadata->getColumnCount();
```

Comment:
This is kind of a strange mix of using `odbc` (native) objects and QGIS wrapper objects. It would be better from my point of view if all of these tasks could be done via the wrapper objects.

Reply:
Done in de55f61.